Automattic/sensei
# Add scripts to build pot file with PHP and JS texts

### Changes proposed in this Pull Request

* Update scripts to generate pot file with the texts from PHP and JS.
* The pot file wasn't updated in this PR.
* It was also introduced a script that adds the reference to the `dist` files in the pot file. The cons about that is for each new dist file, we have to add it to the `additionalReferences` array here (similar to this [WooCommerce script](https://github.com/woocommerce/woocommerce-admin/blob/master/bin/make-i18n-json.sh)): https://github.com/Automattic/sensei/blob/b12c4b9541c84b4994df3b333cde3009874ec0f4/bin/pot-dist-references.js#L8-L13
At some point, we can think some alternative to update this file automatically while building our JS files.

### How it works

1) Clean the temp files.
2) Generate a temporary `pot` file for the js(x) files.
3) Generate the PHP `pot` file, merging the JS `pot` file.
4) Add the dist references as an additional reference to each source occurrence. So the users can translate with the correct reference to the source in http://translate.wordpress.org/.
5) Clean the temp files.

### Testing instructions

If you want to debug the whole process, you can temporarily remove the last `npm run i18n:clean` from the `i18n:build` script in the `package.json`.

* Run `npm run i18n:build`
* Check if the `sensei-lms.pot` was generated correctly.
* Update some texts in the PHP and JS files (for example: `assets/onboarding/index.jsx`), rebuild, and make sure the `sensei-lms.pot` was updated correctly.
* Remove some texts, rebuild, and make sure it was removed from the `sensei-lms.pot` file.

#### Testing the translations running in dev env:

* Update the pot file: `npm run i18n:build`.
* Create a translation `po` file with poedit or running `msginit -l pt_BR -o sensei-lms-pt_BR.po` in the `lang` folder.
* If you generated through the command line, write some translations in the generated file.
* Run `msgfmt -o sensei-lms-pt_BR.mo sensei-lms-pt_BR.po` to generate the mo file.
* Run `wp i18n make-json lang --no-purge` to generate the Jed file for JS messages.
* Move `.mo` and `.json` files to `/wp-content/languages/plugins`.
* Go to WP-admin > Settings > General and update the language.
* Access some JavaScript page (eg: `/wp-admin/admin.php?page=sensei_onboarding`) to test and check if the messages are translated.
* Access some PHP page (eg: `/wp-admin/edit.php?post_type=course`) to test and check if the messages are translated.

### Other observations:

* The documentation wasn't updated yet. It will be done after the PR approval.
  * With this PR, the main script to run the i18n build will be `npm run i18n:build` instead of `gulp pot` (which generates only the PHP part).
* Our i18n build now depends on the `wp i18n make-pot` from the `wp-cli`.
* As a bonus, while running the script `wp i18n make-pot` give us feedback about different translator comments that we can fix in a future PR.

### References:

* The `wp-pot` is restricted only to a PHP parser: https://github.com/wp-pot/wp-pot/blob/master/translation-parser.js#L5
* ~We implemented the same idea than WooCommerce i18n script process~: https://github.com/woocommerce/woocommerce-admin/blob/master/CONTRIBUTING.md#helper-scripts
* Makepot plugin for Babel to generate the po from JS files: https://github.com/WordPress/gutenberg/tree/a1bc317086fca7602f03fcf88ea0c1de514656c3/packages/babel-plugin-makepot
* Implementation details from a core committer, who looks very involved with the i18n part: https://pascalbirchler.com/internationalization-in-wordpress-5-0/
* Demo project: https://github.com/swissspidy/gutenberg-i18n-block
* Core class used for updating/installing language packs (Thanks @jom): https://developer.wordpress.org/reference/classes/language_pack_upgrader/
* More details about JavaScript Internationalization:
  * https://make.wordpress.org/core/2018/05/01/javascript-internationalization-the-missing-pieces/
  * https://developer.wordpress.org/block-editor/developers/internationalization/
  * https://make.wordpress.org/core/2018/11/09/new-javascript-i18n-support-in-wordpress/
